File: bot/telegram.py
from __future__ import annotations

from urllib.parse import quote

from bot.config import TelegramSettings
from bot.http import JsonHttpClient


class TelegramClient:
    def __init__(self, settings: TelegramSettings, http: JsonHttpClient | None = None) -> None:
        self.chat_id = settings.chat_id
        self.endpoint = f"https://api.telegram.org/bot{quote(settings.bot_token, safe=':')}/sendMessage"
        self.http = http or JsonHttpClient()

    def send(self, text: str) -> None:
        self.http.request(
            "POST",
            self.endpoint,
            json_body={
                "chat_id": self.chat_id,
                "text": text[:4_000],
            },
        )

File: supabase/functions/_shared/telegram.ts
import { fetchJson } from "./http.ts";

interface TelegramResponse {
  ok?: boolean;
}

export class TelegramClient {
  constructor(
    private readonly botToken: string,
    private readonly chatId: string,
  ) {}

  async send(text: string): Promise<void> {
    const payload = await fetchJson<TelegramResponse>(
      "Telegram API",
      `https://api.telegram.org/bot${this.botToken}/sendMessage`,
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({ chat_id: this.chatId, text: text.slice(0, 4000) }),
      },
    );
    if (!payload.ok) throw new Error("Telegram API did not confirm the message");
  }
}
